src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +159,163 @@ public virtual byte[] GetValue()
+                        ThrowUnless(asnReadSuccessful);
+                        asnSpan = asnSpan.Slice(bytesConsumed);
+
+                        dirsyncCookie = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
+                        ThrowUnless(asnSpan.Length == bytesConsumed);
+
+                        DirSyncResponseControl dirsync = new DirSyncResponseControl(dirsyncCookie, moreResults != 0, count, controls[i].IsCritical, value);
+                        controls[i] = dirsync;
                     }
-                    else
+                    else if (controls[i].Type == "1.2.840.113556.1.4.474")
                     {
-                        o = BerConverter.Decode("{iie}", value);
-                        Debug.Assert(o != null && o.Length == 3);
-                        position = (int)o[0];
-                        count = (int)o[1];
-                        result = (int)o[2];
+                        // The control is a SortResponseControl. The structure of its value is described as a SortResult in RFC 2891.
+                        ResultCode result;
+                        string attribute = null;
+
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
+                        asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
+
+                        result = AsnDecoder.ReadEnumeratedValue<ResultCode>(asnSpan, AsnEncodingRules.BER, out int bytesConsumed);
+                        asnSpan = asnSpan.Slice(bytesConsumed);
+
+                        // If present, the remaining bytes in the control are expected to be an octet string.
+                        if (!asnSpan.IsEmpty)
+                        {
+                            // Attribute name is optional: AD for example never returns attribute name
+                            scoped Span<byte> attributeNameBuffer;
+
+                            if (asnSpan.Length <= AttributeNameStackAllocationThreshold)
+                            {
+                                asnReadSuccessful = AsnDecoder.TryReadOctetString(asnSpan, attributeNameScratchSpace, AsnEncodingRules.BER, out bytesConsumed, out int octetStringLength);
+                                Debug.Assert(asnReadSuccessful);
+                                attributeNameBuffer = attributeNameScratchSpace.Slice(0, octetStringLength);
+                            }
+                            else
+                            {
+                                attributeNameBuffer = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
+                            }
+                            asnSpan = asnSpan.Slice(bytesConsumed);
+
+                            attribute = s_utf8Encoding.GetString(attributeNameBuffer);
```

Comment:
Since this is using the throw-on-invalid form of the UTF-8 decoder there should be a catch for DecoderFallbackException (probably turning it into a BerConversionException)

Reply:
I think this is also a behavioral change which is virtuous but probably belongs in a breaking change doc.